target/linux/mediatek/patches-6.6/902-arm-add-cmdline-replace.patch

```diff
@@ -0,0 +1,71 @@
+--- a/arch/arm64/Kconfig
++++ b/arch/arm64/Kconfig
+@@ -2316,6 +2316,15 @@ config CMDLINE_OVERRIDE
+ 	  devices. This setting enables using "chosen/cmdline-override" as the
+ 	  cmdline if it exists in the device tree.
+ 
++config CMDLINE_REPLACE
++       bool "Replace specific cmdline from bootloader"
++	default n
++       help
++         Some bootloaders pass specific bootargs to the linux kernel that need
++         to additional behaviors for proper operation. This setting enables
++         replacing bootargs from "chosen/cmdline-replace" at early boot if it
++         exists in the device tree.
++
+ config EFI_STUB
+ 	bool
+ 
+--- a/drivers/of/fdt.c
++++ b/drivers/of/fdt.c
+@@ -1157,6 +1157,11 @@ int __init early_init_dt_scan_chosen(cha
+ 	const void *rng_seed;
+ 	const void *fdt = initial_boot_params;
+ 
++#ifdef CONFIG_CMDLINE_REPLACE
++	char *delim;
++	char *str;
++#endif
++
+ 	node = fdt_path_offset(fdt, "/chosen");
+ 	if (node < 0)
+ 		node = fdt_path_offset(fdt, "/chosen@0");
+@@ -1200,6 +1205,38 @@ int __init early_init_dt_scan_chosen(cha
+ 		strlcpy(cmdline, p, min((int)l, COMMAND_LINE_SIZE));
+ #endif
+ 
++    /* CONFIG_CMDLINE_REPLACE is used to replace a kernel argument at early-
++     * boot with a different device tree option in chosen/bootargs-replace.
++     * This is helpful for boards where u-boot sets bootargs, and is need
++     * to be identified and replaced.
++     */
++#ifdef CONFIG_CMDLINE_REPLACE
++	p = of_get_flat_dt_prop(node, "bootargs-replace", &l);
```

Comment:
`of_get_flat_dt_prop` returns a `const void *`. You are later modifying it, I think this is just a reference to the internal data structure.  I think it would be better to operate on a copy of this string.

Reply:
modified to made a space in front of the cmdline to copy the dt prop so it does not write directly to the `p`.

Reply:
I would prefer if you create a new array on the stack and use that: 
```
char replace[COMMAND_LINE_SIZE];
```

Reply:
the `COMMAND_LINE_SIZE` is `2048` on arm64, and the compiler complain with these error and cannot successfully build:
 ```
 warning: the frame size of 2064 bytes is larger than 2048 bytes [-Wframe-larger-than]
 ```

Reply:
could I create a new array with a smaller length like this?
```
char replace[COMMAND_LINE_SIZE / 2];
```

Reply:
yes half the size should be fine.

Reply:
modified to use a new stack array `replace`.